weave/index/index_sql.py

```diff
@@ -0,0 +1,869 @@
+"""Wherein is contained the concrete SQLite implementation of the Index."""
+# Pylint doesn't like the similarity between this file and the SQLite file, but
+# it doesn't make sense to write shared functions for them. So ignore pylint.
+# pylint: disable=duplicate-code
+import json
+import os
+import warnings
+from datetime import datetime
+
+import ast
+import dateutil
+import pandas as pd
+# Try-Except required to make pyodbc an optional dependency.
+try:
+    import pyodbc
+except ImportError:
+    _HAS_PYODBC = False
+else:
+    _HAS_PYODBC = True
+
+from ..config import get_index_column_names
+from .index_abc import IndexABC
+from .list_baskets import _get_list_of_basket_jsons
+from .validate_basket import validate_basket_dict
+
+class IndexSQL(IndexABC):
+    """Concrete implementation of Index, using SQL."""
+    def __init__(self, file_system, pantry_path, **kwargs):
+        """Initializes the Index class.
+
+        Parameters
+        ----------
+        file_system: fsspec object
+            The fsspec object which hosts the pantry we desire to index.
+        pantry_path: str
+            Path to the pantry root which we want to index.
+        **database_name: str (default='weave_db')
+            DB to be used. If none is set, defaults to 'weave_db'.
+        **encrypt: bool (default=False)
+            Whether or not to encrypt the database (disabled by default, due
+            to the fact that it requires additional configuration).
+        **odbc_driver: str (default='{ODBC Driver 18 for SQL Server}')
+            Driver to be used for the connection. Usually similar to:
+            '{ODBC Driver <18, 17, 13...> for SQL Server}'.
+        **pantry_schema: str (default=<pantry_path>)
+            The schema to use for the pantry. If none is set, defaults to the
+            pantry path (with _ replacements when necessary).
+        """
+        if not _HAS_PYODBC:
+            raise ImportError("Missing Dependency. The package 'pyodbc' "
+                              "is required to use this class")
+
+        # Check that the required environment variables are set.
+        try:
+            # We want to fail early if these are not set, so we check them here
+            # even though they are not used until we connect.
+            # Pylint thinks this is pointless, so pylint is getting ignored.
+            # pylint: disable=pointless-statement
+            os.environ["MSSQL_HOST"]
+            os.environ["MSSQL_USERNAME"]
+            os.environ["MSSQL_PASSWORD"]
+            # pylint: enable=pointless-statement
+        except KeyError as key_error:
+            raise KeyError("The following environment variables must be set "
+                           "to use this class: MSSQL_HOST, MSSQL_USERNAME, "
+                           "MSSQL_PASSWORD.") from key_error
```

Comment:
In the pytest for sql it doesn't look like this section of code is being tested.